File: web/src/app/admin/fees/fee-payment.ts
import { PaymentMethod } from './payment-method';

export interface FeePayment {
  id?: number;
  createdAt?: string | null;
  schoolId?: number | null;
  studentId: number | null;
  studentName?: string | null;
  studentNumber?: string | null;
  feeStructureId: number | null;
  gradeId?: number | null;
  gradeName?: string | null;
  term?: string | null;
  academicYear?: string | null;
  amount: number | null;
  paymentDate: string;
  paymentMethod: PaymentMethod | null;
  referenceNumber?: string | null;
  notes?: string | null;
  totalFee?: number | null;
  totalPaid?: number | null;
  balance?: number | null;
  reversed?: boolean;
  reversedAt?: string | null;
  reversalReason?: string | null;
}

File: api/src/main/java/com/tsoinyane/api/fee/FeePayment.java
package com.tsoinyane.api.fee;

import com.tsoinyane.api.common.BaseEntity;
import com.tsoinyane.api.student.Student;
import com.tsoinyane.api.user.User;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.FetchType;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import jakarta.persistence.Table;
import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;
import lombok.ToString;
import lombok.experimental.SuperBuilder;

import java.time.Instant;
import java.time.LocalDate;

@Data
@NoArgsConstructor
@AllArgsConstructor
@SuperBuilder
@Entity
@Table(name = "fee_payment")
@EqualsAndHashCode(callSuper = true, exclude = {"student", "feeStructure"})
@ToString(callSuper = true, exclude = {"student", "feeStructure"})
public class FeePayment extends BaseEntity {

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "student_id", nullable = false)
    private Student student;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "fee_structure_id", nullable = false)
    private FeeStructure feeStructure;

    @Column(nullable = false)
    private Double amount;

    @Column(name = "payment_date", nullable = false)
    private LocalDate paymentDate;

    @Enumerated(EnumType.STRING)
    @Column(name = "payment_method", nullable = false, length = 30)
    private PaymentMethod paymentMethod;

    @Column(name = "reference_number", length = 100)
    private String referenceNumber;

    @Column(length = 500)
    private String notes;

    @Column(nullable = false)
    private Boolean reversed = false;

    @Column(name = "reversed_at")
    private Instant reversedAt;

    @Column(name = "reversal_reason", length = 500)
    private String reversalReason;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "reversed_by_id")
    private User reversedBy;
}
